sklearn/ensemble/_hist_gradient_boosting/gradient_boosting.py

```diff
@@ -0,0 +1,833 @@
+
+            # maybe we could also early stop if all the trees are stumps?
+            if should_early_stop:
+                break
+
+        if self.verbose:
+            duration = time() - fit_start_time
+            n_total_leaves = sum(
+                predictor.get_n_leaf_nodes()
+                for predictors_at_ith_iteration in self._predictors
+                for predictor in predictors_at_ith_iteration)
+            n_predictors = sum(
+                len(predictors_at_ith_iteration)
+                for predictors_at_ith_iteration in self._predictors)
+            print("Fit {} trees in {:.3f} s, ({} total leaves)".format(
+                n_predictors, duration, n_total_leaves))
+            print("{:<32} {:.3f}s".format('Time spent computing histograms:',
+                                          acc_compute_hist_time))
+            print("{:<32} {:.3f}s".format('Time spent finding best splits:',
+                                          acc_find_split_time))
+            print("{:<32} {:.3f}s".format('Time spent applying splits:',
+                                          acc_apply_split_time))
+            print("{:<32} {:.3f}s".format('Time spent predicting:',
+                                          acc_prediction_time))
+
+        self.train_score_ = np.asarray(self.train_score_)
+        self.validation_score_ = np.asarray(self.validation_score_)
+        del self._in_fit  # hard delete so we're sure it can't be used anymore
+        return self
+
+    def _check_early_stopping_scorer(self, X_binned_small_train, y_small_train,
+                                     X_binned_val, y_val):
+        """Check if fitting should be early-stopped based on scorer.
+
+        Scores are computed on validation data or on training data.
+        """
+
+        self.train_score_.append(
+            self.scorer_(self, X_binned_small_train, y_small_train)
+        )
+
+        if self._use_validation_data:
+            self.validation_score_.append(
+                self.scorer_(self, X_binned_val, y_val)
+            )
+            return self._should_stop(self.validation_score_)
+        else:
+            return self._should_stop(self.train_score_)
+
+    def _check_early_stopping_loss(self,
+                                   raw_predictions,
+                                   y_train,
+                                   raw_predictions_val,
+                                   y_val):
+        """Check if fitting should be early-stopped based on loss.
+
+        Scores are computed on validation data or on training data.
+        """
+
+        self.train_score_.append(
+            -self.loss_(y_train, raw_predictions)
+        )
+
+        if self._use_validation_data:
+            self.validation_score_.append(
+                -self.loss_(y_val, raw_predictions_val)
+            )
+            return self._should_stop(self.validation_score_)
+        else:
```

Comment:
flake8 would suggest to remove the else with the indent and just put the return

Reply:
As far as i can tell pep8 has no recommendation regarding this. Again, I find


```py
if something:
   return foo
else:
   return bar
```

cleaner and explicit.

Reading

```py
if something:
   return foo

return bar
```

makes me think that the function always returns `bar` at first glance. I try to only use it for catching errors and such.

Reply:
Actually, that's pylint which was telling me something in vscode: SoCo/SoCo#500